File: ruby/cluster_management/lib/update.rb
require 'aws-sdk-core'
require 'aws-sdk-dsql'

def update_cluster(region, identifier)
    begin
      # Create client with default configuration and credentials
      client = Aws::DSQL::Client.new(region: region)

      update_response = client.update_cluster(
          identifier: identifier,
          deletion_protection_enabled: false
      )

      client.tag_resource(
          resource_arn: update_response.arn,
          tags: {
              "Function" => "Billing",
              "Environment" => "Production"
          }
      )
      raise "Unexpected status when updating cluster: #{update_response.status}" unless update_response.status == 'UPDATING'
      update_response
    rescue Aws::Errors::ServiceError => e
      raise "Failed to update cluster details: #{e.message}"
    end
end

File: rust/cluster_management/src/update.rs
use aws_config::load_defaults;
use aws_sdk_dsql::{config::{BehaviorVersion, Region}, Client, Config};
use aws_sdk_dsql::operation::update_cluster::UpdateClusterOutput;

/// Create a client. We will use this later for performing operations on the cluster.
async fn dsql_client(region: &'static str) -> Client {
    // Load default SDK configuration
    let sdk_defaults = load_defaults(BehaviorVersion::latest()).await;

    // You can set your own credentials by following this guide
    // <https://docs.aws.amazon.com/sdk-for-rust/latest/dg/credproviders.html>
    let credentials = sdk_defaults
        .credentials_provider()
        .unwrap();

    let config = Config::builder()
        .behavior_version(BehaviorVersion::latest())
        .credentials_provider(credentials)
        .region(Region::new(region))
        .build();
    
    Client::from_conf(config)
}

// Update a DSQL cluster and set delete protection to false. Also add new tags.
pub async fn update_cluster(region: &'static str, identifier: String) -> UpdateClusterOutput {
    let client = dsql_client(region).await;
    // Update delete protection
    let update_response = client
        .update_cluster()
        .identifier(identifier)
        .deletion_protection_enabled(false)
        .send()
        .await
        .unwrap();

    // Add new tags
    client
        .tag_resource()
        .resource_arn(update_response.arn().to_owned())
        .tags(String::from("Function"), String::from("Billing"))
        .tags(String::from("Environment"), String::from("Production"))
        .send()
        .await
        .unwrap();

    update_response
}

#[tokio::main(flavor = "current_thread")]
#[allow(dead_code)]
pub async fn main() -> anyhow::Result<()> {
    let region = "us-east-1";

    update_cluster(region, "<your cluster id>".to_owned()).await;

    Ok(())
}
